File: cloud/src/main/java/server/model/Product.java
package server.model;

import org.json.JSONObject;
import server.model.crdts.CCounter;

public class Product {

    public String uid;
    public String name;
    public CCounter counter;

    /**
     *
     * @param {String} name - Product name
     * @param {String} u_id - User identifier
     */
    public Product(String name, String u_id) {
        this.uid = u_id;
        this.name = name;
        this.counter = new CCounter(u_id);
    }

    /**
     * Get product name
     *
     * @return Product name
     */
    public String getName() {
        return name;
    }

    /**
     * Get product amount
     *
     * @return Product amount
     */
    public CCounter getCounter() {
        return counter;
    }

    public void setID(String u_id) {
        this.uid = u_id;
        this.counter.setID(u_id);
    }

    /**
     * Get product amount
     *
     * @return  {Number} Product amount
     */
    public int value() {
        return this.counter.read();
    }

    /**
     * Increment product amount
     *
     * @param toSum Amount to increment
     */
    public void inc(int toSum) {
        counter.inc(toSum);
    }

    /**
     * Decrement product amount
     *
     * @param toSum Amount to decrement
     */
    public void dec(int toSum) {
        counter.dec(toSum);
    }

    /**
     * Merge two products
     *
     * @param product Product to merge
     */
    public void join(Product product) {
        counter.join(product.counter);
    }

    /**
     * Create a Product from a JSON object
     *
     * @param json JSON object in string format
     */
    public void fromJSON(JSONObject json) {
        /* test: { \"name\" : \"ShopList\", \"counter\" : \"18\" } */

        this.name = json.getString("name");
        this.counter = new CCounter();
        this.counter.fromJSON((JSONObject) json.get("counter"));
    }

    public JSONObject toJSON() {
        return this.counter.toJSON();
    }
}

File: local/src/lib/product.js
import { CCounter } from "$lib/crdt/ccounter";
import { PNCounter } from "$lib/crdt/pncounter";

export class Product {
    #name = null;
    #counter = null;
    #uid = null;

    /**
     *
     * @param {String} name - Product name
     * @param {Strinh} u_id - User identifier
     */
    constructor(name, u_id) {
        this.#uid = u_id;
        this.#name = name;
        this.#counter = new CCounter(u_id);
    }

    setUUID(u_id) {
        this.#uid = u_id;
        this.#counter.setID(u_id);
    }

    /**
     * Get product name
     *
     * @returns {String} Product name
     */
    getName() {
        return this.#name;
    }

    /**
     * Get product counter
     *
     * @returns {PNCounter} Product counter
     */
    getCounter() {
        const res = this.#counter;
        return res;
    }

    /**
     * Get product amount
     *
     * @returns {Number} Product amount
     */
    value() {
        return this.#counter.read();
    }

    /**
     * Increment product amount
     *
     * @param {Number} tosum
     */
    inc(tosum = 1) {
        this.#counter.inc(tosum);
    }

    /**
     * Decrement product amount
     *
     * @param {Number} tosum
     */
    dec(tosum = 1) {
        this.#counter.dec(tosum);
    }

    /**
     * Merge two products
     *
     * @param {Product} product
     */
    join(product) {
        this.#counter.join(product.#counter);
    }

    /**
     * Create a Product from a JSON object
     *
     * @param {JSON} product
     */
    fromJSON(product) {
        this.#name = product.name;
        this.#counter = new CCounter(this.#uid);
        this.#counter.fromJSON(product.counter);
    }

    toJSON() {
        const res = {
            counter: this.#counter.toJSON(),
        };

        return res;
    }
}
